modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/js/LanguageSelector.js

```diff
@@ -0,0 +1,143 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayDropDown from '@clayui/drop-down';
+import ClayIcon from '@clayui/icon';
+import PropTypes from 'prop-types';
+import React, {useEffect, useState} from 'react';
+
+function getLanguage(id) {
+	const text = id.replace('_', '-');
+	const icon = text.toLowerCase();
+
+	return {
+		icon,
+		text,
+	};
+}
+
+function LanguageSelector({languageIds, onChange, selectedLanguageId}) {
+	const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+	const language = getLanguage(selectedLanguageId);
+
+	return (
+		<ClayDropDown
+			active={isDropdownOpen}
+			onActiveChange={setIsDropdownOpen}
+			trigger={
+				<ClayButton displayType="secondary" monospaced>
+					<span className="inline-item">
+						<ClayIcon symbol={language.icon} />
+					</span>
+					<span className="btn-section">{language.text}</span>
+				</ClayButton>
+			}
+		>
+			<ClayDropDown.ItemList>
+				{languageIds.map((id) => {
+					const {icon, text} = getLanguage(id);
+
+					return (
+						<ClayDropDown.Item
+							active={id === selectedLanguageId}
+							key={id}
+							onClick={() => {
+								onChange(id);
+								setIsDropdownOpen(false);
+							}}
+						>
+							<span className="inline-item inline-item-before">
+								<ClayIcon symbol={icon} />
+							</span>
+							{text}{' '}
+						</ClayDropDown.Item>
+					);
+				})}
+			</ClayDropDown.ItemList>
+		</ClayDropDown>
+	);
+}
+
+LanguageSelector.propTypes = {
+	languageIds: PropTypes.arrayOf(PropTypes.string).isRequired,
+	onChange: PropTypes.func.isRequired,
+	selectedLanguageId: PropTypes.string.isRequired,
+};
+
+function addAviableLocale(localeId, portletNamespace) {
+	const availableLocalesInput = document.getElementById(
+		`${portletNamespace}availableLocales`
+	);
+
+	if (availableLocalesInput) {
+		const availableLocales = availableLocalesInput.value
+			? availableLocalesInput.value.split(',')
+			: [];
+
+		availableLocalesInput.value = [
+			...new Set([...availableLocales, localeId]),
+		].join(',');
+	}
+}
+
+function DataEngineLanguageSelector({
+	ddmStructureIds,
+	portletNamespace,
+	selectedLanguageId: initialSelectedLanguageId,
+	...restProps
+}) {
+	const [selectedLanguageId, setSelectedLanguageId] = useState(
+		initialSelectedLanguageId
+	);
+
+	const handleLocaleChange = (localeId) => {
+		setSelectedLanguageId(localeId);
+		addAviableLocale(localeId, portletNamespace);
+	};
+
+	useEffect(() => {
+		ddmStructureIds.forEach((ddmStructureId) => {
```

Comment:
how does this works? Is the `dataEngineLayoutRenderer` component instantiated every time `selectedLanguageId` changes? Sounds weird if so. In this case we're running this code every time `portletNamespace` ,  `selectedLanguageId` and `ddmStructureIds` changes, but I guess we only want to run it if `ddmStrucureIds` change.

Reply:
No, we want to run if `selectedLanguageId ` changes, but the other two are props if some change we want to run too.